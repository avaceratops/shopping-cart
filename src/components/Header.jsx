import { NavLink } from 'react-router-dom';
import CartButton from './CartButton';
import '../styles/Header.scss';

export default function Header({ cart }) {
  return (
    <header className="px-5">
      <div className="border-b border-gray-200">
        <nav className="flex h-16 items-center justify-between text-gray-700">
          <section className="flex-1">
            <h1 className="text-2xl font-bold leading-none tracking-tight text-black">
              shopping cart
            </h1>
          </section>
          <ul className="flex h-full space-x-8 text-sm font-medium">
            <NavLink to="/">
              <button>Home</button>
            </NavLink>
            <NavLink to="/warhammer-40k">
              <button>Warhammer 40,000</button>
            </NavLink>
            <NavLink to="/age-of-sigmar">
              <button>Age of Sigmar</button>
            </NavLink>
          </ul>
          <section className="flex flex-1 justify-end">
            <CartButton cart={cart} />
          </section>
        </nav>
      </div>
    </header>
  );
}
