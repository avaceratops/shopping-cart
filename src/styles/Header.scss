nav a {
  @apply border-b-2 border-transparent;

  &:hover {
    @apply text-gray-800;
  }

  &.active {
    @apply border-indigo-600 text-indigo-600;

    &:hover {
      @apply text-indigo-700;
    }
  }

  button {
    @apply h-full;
  }
}
