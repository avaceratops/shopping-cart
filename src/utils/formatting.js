export function formatPrice(pence) {
  const pounds = pence / 100;
  return pounds.toLocaleString('en-GB', { style: 'currency', currency: 'GBP' });
}
