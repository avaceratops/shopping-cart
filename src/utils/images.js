export const GAME_LOGOS = [
  '/images/logos/https___trade.games-workshop.com_assets_2020_05_Warhammer_40K_Logo',
  '/images/logos/https___trade.games-workshop.com_assets_2019_05_WHAoSLogoNEW2018',
];

export const PROMOTIONAL_IMAGES = [
  '/images/promotional/https___trade.games-workshop.com_assets_2023_12_Trade_Header_30-12_leviathan',
  '/images/promotional/https___trade.games-workshop.com_assets_2023_07_15-07_Trade_Header_Librarian',
  '/images/promotional/https___trade.games-workshop.com_assets_2023_05_Trade_Header_13-05',
  '/images/promotional/https___trade.games-workshop.com_assets_2023_04_Trade_Header_08-04',
];
