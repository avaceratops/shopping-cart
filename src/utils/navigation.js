export const NAVIGATION_LINKS = [
  { name: 'Home', to: '/' },
  { name: 'Warhammer 40,000', to: '/warhammer-40k' },
  { name: 'Age of Sigmar', to: '/age-of-sigmar' },
];
