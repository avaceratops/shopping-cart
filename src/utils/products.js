export const MAX_PURCHASE_QUANTITY = 8;

export const FACTION_NAMES = {
  'warhammer-40k': [
    'Aeldari',
    'Astra Militarum',
    'Blood Angels',
    'Dark Angels',
    'Necrons',
    'Orks',
    "T'au Empire",
    'Thousand Sons',
    'Tyranids',
    'World Eaters',
  ],
  'age-of-sigmar': [
    'Nighthaunt',
    'Orruk Warclans',
    'Ossiarch Bonereapers',
    'Seraphon',
    'Skaven',
    'Slaves to Darkness',
    'Sons of Behemat',
    'Sylvaneth',
  ],
};

export const GAME_DESCRIPTIONS = {
  'warhammer-40k':
    "Embark on an intergalactic odyssey with Warhammer 40k at Shopping Cart. Whether you're a seasoned warrior or a newcomer, our extensive array of miniatures, rulebooks, and accessories promises an immersive gaming experience like no other. From the mighty Space Marines to the cunning Orks and beyond, gather your forces and brace yourself for epic tabletop battles.\n\nTake advantage of competitive prices that allow you to bolster your armies without compromise. Need help assembling your Warhammer 40k army? Our knowledgeable staff is on hand to offer assistance every step of the way. And with free delivery on all orders over £50, conquering the galaxy has never been more convenient.",
  'age-of-sigmar':
    'Plunge into an epic journey through Age of Sigmar at Shopping Cart. Discover the vast lore and captivating world of this legendary tabletop game, as you command your armies across the Mortal Realms. From the stalwart Stormcast Eternals to the cunning Orruks and beyond, muster your legions and ready yourself for grand skirmishes.\n\nTake advantage of competitive prices to expand your collection and conquer the Mortal Realms without breaking the bank. If you need assistance selecting the perfect miniatures, our knowledgeable staff are here to help. And with free delivery on all orders over £50, building the perfect army has never been easier.',
};

export function getGameByProductId(id) {
  const start = id.charAt(0);
  return start === '0' ? 'warhammer-40k' : start === '1' ? 'age-of-sigmar' : null;
}

export function sortByAvailability(a, b) {
  if (a.stock === 0 && b.stock > 0) {
    return 1;
  } else if (a.stock > 0 && b.stock === 0) {
    return -1;
  }
  return 0;
}

function sortByBestSelling(a, b) {
  return b.sold - a.sold;
}

function sortByNameAsc(a, b) {
  return a.name > b.name ? 1 : b.name > a.name ? -1 : 0;
}

function sortByNameDes(a, b) {
  return b.name > a.name ? 1 : a.name > b.name ? -1 : 0;
}

function sortByPriceAsc(a, b) {
  return a.price - b.price;
}

function sortByPriceDes(a, b) {
  return b.price - a.price;
}

export const SORT_METHODS = {
  'Best Selling': sortByBestSelling,
  'Alphabetically, A-Z': sortByNameAsc,
  'Alphabetically, Z-A': sortByNameDes,
  'Price, low to high': sortByPriceAsc,
  'Price, high to low': sortByPriceDes,
};

export function getSortFunction(sortMethod) {
  const sortFunction = SORT_METHODS[sortMethod];
  if (!sortFunction) {
    throw new Error('Invalid sort method specified');
  }
  return sortFunction;
}
