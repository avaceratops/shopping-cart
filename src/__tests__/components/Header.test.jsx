import { vi } from 'vitest';
import { BrowserRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import { NAVIGATION_LINKS } from '../../utils/navigation';
import Header from '../../components/Header';

const updateCartItem = vi.fn();
const removeFromCart = vi.fn();
const clearCart = vi.fn();

it('renders header correctly', () => {
  const { container } = render(
    <BrowserRouter>
      <Header
        cart={[]}
        updateCartItem={updateCartItem}
        removeFromCart={removeFromCart}
        clearCart={clearCart}
      />
    </BrowserRouter>
  );

  expect(container).toMatchSnapshot();
});

it('renders all navigation links correctly', () => {
  render(
    <BrowserRouter>
      <Header
        cart={[]}
        updateCartItem={updateCartItem}
        removeFromCart={removeFromCart}
        clearCart={clearCart}
      />
    </BrowserRouter>
  );

  NAVIGATION_LINKS.filter((item) => item.name !== 'Home').forEach((item) => {
    expect(screen.getByRole('link', { name: item.name })).toBeInTheDocument();
  });
});
