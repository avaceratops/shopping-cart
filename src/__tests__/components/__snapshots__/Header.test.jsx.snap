// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders header correctly 1`] = `
<div>
  <header
    class="sticky top-0 z-10 w-full max-w-screen-2xl bg-white/90 px-5 backdrop-blur"
  >
    <div
      class="border-b border-gray-200"
    >
      <nav
        class="flex h-16 items-center justify-between text-gray-700"
      >
        <section
          class="flex flex-1 md:hidden"
        >
          <button
            class="rounded-md p-2 text-gray-400 hover:bg-gray-400 hover:text-white"
          >
            <span
              class="sr-only"
            >
              Open Menu
            </span>
            <svg
              aria-hidden="true"
              class="block h-6 w-6"
              data-slot="icon"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
        </section>
        <section
          class="hidden xs:block lg:flex-1"
        >
          <h1
            class="text-2xl font-bold leading-none tracking-tight text-black"
          >
            <a
              aria-current="page"
              class="active"
              href="/"
            >
              shopping cart
            </a>
          </h1>
        </section>
        <ul
          class="hidden h-full space-x-8 text-sm font-medium md:flex"
        >
          <li>
            <a
              class="_navLink_800479"
              href="/warhammer-40k"
            >
              Warhammer 40,000
            </a>
          </li>
          <li>
            <a
              class="_navLink_800479"
              href="/age-of-sigmar"
            >
              Age of Sigmar
            </a>
          </li>
        </ul>
        <section
          class="flex flex-1 justify-end md:flex-initial lg:flex-1"
        >
          <button
            class="group flex items-center rounded-md p-2 hover:bg-gray-400 hover:text-white"
          >
            <span
              class="sr-only"
            >
              Cart
            </span>
            <svg
              aria-hidden="true"
              class="block h-6 w-6 flex-shrink-0 text-gray-400 group-hover:text-white"
              data-slot="icon"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.75 10.5V6a3.75 3.75 0 1 0-7.5 0v4.5m11.356-1.993 1.263 12c.07.665-.45 1.243-1.119 1.243H4.25a1.125 1.125 0 0 1-1.12-1.243l1.264-12A1.125 1.125 0 0 1 5.513 7.5h12.974c.576 0 1.059.435 1.119 1.007ZM8.625 10.5a.375.375 0 1 1-.75 0 .375.375 0 0 1 .75 0Zm7.5 0a.375.375 0 1 1-.75 0 .375.375 0 0 1 .75 0Z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            <span
              class="ml-2 text-sm font-medium text-gray-800 group-hover:text-white"
            >
              0
            </span>
          </button>
        </section>
      </nav>
    </div>
  </header>
</div>
`;
